File: src/bootstrapper/ecosphere/objects/interface.py
import ecosphere.objects.base


class EcoInterfaceEntry:

    def set_return_type(self, value):
        self._return_type = value
    
    def set_arg_type(self, index, value):
        self._arg_types[index] = value

    def serialize(self, serializer):
        serializer.write_object(self._name)
        if self._return_type is None:
            serializer.write_message('ecosphere.object.interface.default')
        else:
            serializer.write_object(self._return_type)
        if self._has_varargs:
            serializer.write_vlq(1)
        else:
            serializer.write_vlq(0)
        serializer.write_vlq(len(self._arg_types))
        for at in self._arg_types:
            if at is None:
                serializer.write_message('ecosphere.object.interface.default')
            else:
                serializer.write_object(at)

    def __init__(self, name, arg_count, has_varargs):
        self._return_type = None
        self._name = name
        self._arg_types = [None] * arg_count
        self._has_varargs = has_varargs


class EcoInterface(ecosphere.objects.base.EcoObject):
    
    def add_entry(self, entry):
        self._entries.append(entry)
    
    def serialize(self, serializer):
        if not serializer.try_serialize_known_object(self):
            serializer.write_message('ecosphere.object.interface')
            serializer.write_vlq(serializer.add_object(self))
            serializer.write_vlq(len(self._entries))
            for entry in self._entries:
                entry.serialize(serializer)

    def __init__(self):
        super().__init__
        self._entries = list()

File: src/ecore/io/econnect/reader/builtins/objects/interface.c
#include "interface.h"

#include <ecore/io/econnect/reader/parser.h>
#include <ecore/objects/vm/interface/interface.h>
#include <ecore/vm/memory/memory.h>


bool Eco_EConnect_Builtin_GetInterface(struct Eco_EConnect_Reader* reader,
                                       struct Eco_EConnect_Result* result)
{
    unsigned int               id;
    unsigned int               parent_count;
    unsigned int               entry_count;
    unsigned int               index;
    unsigned int               arg;
    struct Eco_Interface*      the_interface;
    struct Eco_Object*         obj;
    struct Eco_InterfaceEntry  entry;

    id            = Eco_EConnect_ParseUInt(&reader->stream);
    parent_count  = Eco_EConnect_ParseUInt(&reader->stream);
    entry_count   = Eco_EConnect_ParseUInt(&reader->stream);
    the_interface = Eco_Interface_New(parent_count, entry_count);

    Eco_EConnect_Instance_OptionallyBindObject(reader->instance, (struct Eco_Object*) the_interface, id);

    for (index = 0; index < parent_count; index++)
    {
        if (!Eco_EConnect_Reader_ReadObject(reader, result, &obj))
            return false;
        the_interface->parents[index] = (struct Eco_Interface*) obj;
    }

    for (index = 0; index < entry_count; index++)
    {
        if (!Eco_EConnect_Reader_ReadObject(reader, result, (struct Eco_Object**) &entry.key))
            return false;
        if (!Eco_EConnect_Reader_ReadObject(reader, result, (struct Eco_Object**) &entry.return_type))
            return false;
        entry.has_varargs = Eco_EConnect_ParseUInt(&reader->stream) != 0;
        entry.arg_count = Eco_EConnect_ParseUInt(&reader->stream);
        for (arg = 0; arg < entry.arg_count; arg++)
        {
            if (!Eco_EConnect_Reader_ReadObject(reader, result, (struct Eco_Object**) &entry.arg_types[arg]))
                return false;
        }
        the_interface->entries[index] = entry;
    }

    Eco_EConnect_Result_Create_Object(result, (struct Eco_Object*) the_interface);

    return true;
}

bool Eco_EConnect_Builtin_GetDefaultInterface(struct Eco_EConnect_Reader* reader,
                                              struct Eco_EConnect_Result* result)
{
    Eco_EConnect_Result_Create_Object(result, (struct Eco_Object*) Eco_Interface_GetDefaultInterface());
    return true;
}
